### ISSUE TITLE
(Batched) Matrix Operations are much slower on GPU

### ISSUE THREAD

**Issue by ssydasheng** [google/jax#5023]
Hi, 

I am profiling the JAX operations for batched matrix operations on GPU and CPU. It is very surprising that GPU is much slower compared to CPU. For the code below, when **B=1**, the printing time is **CPU= 0.00016665 vs GPU= 0.0067.** Furthermore, when **B=3**, the printing time is **CPU=0.00016355514 vs GPU=0.01804113**. Therefore, GPU is not only lower than CPU, it cannot enjoy the parallel benefits for batching operations either. 

For the platform, I am using 2-CPUs, GPU=P100, jax=0.2.1, jaxlib=0.1.55. I wonder if this is a known issue, and how can I fix this ? I am particularly interested in how to exploit the parallel benefits for batching operations. 

```
import time
from jax import jit, random
import jax.numpy as jnp

B, N, N = 3, 2048, 2048

@jit
def run(iter):
    rng=random.PRNGKey(iter)
    a = random.normal(rng, shape=[B, N, N])
    a = a @ jnp.transpose(a, [0, 2, 1])
    b = jnp.linalg.cholesky(a)
    return b


for _ in range(20):
    run(_)

start = time.time()
run(21)
print(time.time() - start)
```

**Comment by JossWhittle** [comment]
Hi I think I've also been hitting a similar problem of silent recompilations that eat up a lot of time. Try enabling JAX to log every time it triggers a recompilation and check that it is triggering only once for you on the first iteration. i.e.

```
JAX_LOG_COMPILES=1 python test.py
```

There was a bit in the documentation I read about "lists not being hashable, while tuples are" so potentially the way that you pass lists to `random.normal` and for `shape` and to `jnp.transpose` for the axis order are causing those functions to always be compiled every iteration. Changing these to tuples using the `(,)` syntax may fix this if it is silently recompiling.

**Comment by ssydasheng** [comment]
Hi thanks for the suggestion. I tried adding `os.environ["JAX_LOG_COMPILES"] = "1"`, both CPU and GPU output the compilations signal only once. I also tried changing the lists to tuples, but it made no differences either.

There is something more strange. I replaced `run(21)` with `for _ in range(21, 100): run(_)`. When **B=1**, the printing time is  **CPU=21.4955 vs GPU=0.5453**.  When **B=3**, the printing time is **CPU=70.6205 vs GPU=1.4233**.

**Comment by JossWhittle** [comment]
I wonder if the computation has actually finished at the time it returns to python control flow. Can you try changing the invocation to be `run(_).block_until_ready()` as suggested here https://jax.readthedocs.io/en/latest/async_dispatch.html

**Comment by ssydasheng** [comment]
You are exactly right ! After I set `run(_).block_until_ready()`, when **B=1**, the printing time is **CPU=24.23 vs GPU=0.544**; when **B=3**, the printing time is **CPU=80.69 vs GPU=1.43**. 

A final question, is it possible to parallel these operations ? Ideally I expect the GPU should be faster by parallelisms when **B>1** . However, for **B=1,3,10**, the time is **0.544, 1.43, 3.918**, which I think not much a speed up.

**Comment by hawkinsp** [comment]
Currently on GPU we do use batched algorithms for some matrix operations (e.g. Cholesky decomposition), but only for small matrices. Ultimately we're just calling Cusolver here at the moment.

On CPU more parallelism is certainly possible, although it seems your interest is on GPU.

In principle on GPU we could run multiple operations in parallel (using multiple GPU streams), although it's not something we support at the moment. It's also not obvious to me whether it would be profitable for a 2048x2048 matrix, although I don't know without trying it.

I expect support for multiple GPU streams is likely to be something we can experiment with more seriously in six or so months after some runtime and compiler changes have landed.

**Comment by ssydasheng** [comment]
Thanks for your clarification !

**Comment by ssydasheng** [comment]
Thank you again @hawkinsp for the explanations. I wonder if you have any suggestions for me to speed up the batch Cholesky decompositions (~1000x1000). Cholesky decompositions take quadratic memory usage, thus I reckon the GPU memory will be sufficient if the batch is not quite large. Why do you think there might not be benefits from parallelisms ?  

To speed it up, it seems that `pmap` is one potential approach when multi-GPUs are available. If only one GPU is available to me, is the `jax.numpy.cholesky` already the fastest, or there are some workarounds ? Any suggestions will be very helpful, thanks.

**Comment by nouiz** [comment]
We do not think of any war around for single GPU.
Any objection that we close this issue?
